Update Prompts UI (enhanced with client last used action sorting)

vovakulikov opened the pull request:
Part of https://linear.app/sourcegraph/issue/SRCH-1103/update-prompts-ui-welcome-screen

This PR does a few things (TL;DR I suggest checking the [loom](https://www.loom.com/share/3922d57405fc4db2b328a317c211231e?sid=28e903cd-aa63-4c7c-93cb-262b1de6ef7d))
 
- Improve prompts and commands list UI (you can find the latest designs for it [here](https://www.figma.com/design/f078wFMKsIOaEwj7Iwj5xy/Unified-Cody?node-id=4845-20933&node-type=symbol&t=Ao2Y68OYj5lFV3rA-0))
- Adds smart last-used prompts/commands sorting

<img width="1136" alt="Screenshot 2024-10-02 at 02 52 01" src="https://github.com/user-attachments/assets/2d4f2991-6086-4e1b-8b49-10b85c5c35e0">

So this PR is another attempt to bring commands and prompts closer and smooth transition from commands to prompts, so in this PR
- There is no separation between commands and prompts it is basically one list
- on the welcome screen we show only the last 4 used actions (action here means prompts or commands)
- By default, we show out-of-the-box actions 
- If we have custom commands or custom prompts we shift the list to show the last created prompts or commands
- If you used some prompts or commands it will show up on top of the list on the welcome screen 

This is a crucial improvement for customers with a possible high count of prompts because it helps with discoverability and we don't bombard the welcome screen with a big set of items in the suggestion list

## Future follow-ups
- In designs, we have a Search Prompt button below the action list on the welcome screen, for simplicity now it's just a link to the prompt tab but in future, we should open in place modal UI with a search prompt UI 
The last used action is currently implemented with local storage, later we should track usage on the backend  

## Test plan
- Check that you can see only 4 items max on the welcome screen 
- Check that you can find prompts that you have on prompts tab or prompts toolbar 
- Check that last used prompts/commands can be found on the welcome screen right after you used them

dominiccooney commented on lib/shared/src/misc/rpc/webviewAPI.ts:
```diff
@@ -114,27 +114,42 @@ export interface MentionMenuData {
+    //  */
+    // commands: CodyCommand[]
+
+    isPromptsSupported: boolean
```
This is an awkward name because "is" is singular but "prompts" is plural. Consider `supportsPrompts`, or `arePromptsSupported`, or if we really want "is" then `isPromptSupporting` I guess...

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -11,85 +12,75 @@ import {
 import { FIXTURE_COMMANDS } from '../../webviews/components/promptList/fixtures'
 import { getCodyCommandList } from '../commands/CommandsController'
 
+/** For testing only. */
```
Is this for *testing* per se, or development?

vovakulikov replied:
I actually have no idea, this code was here before me from the very beginning of commands/prompts functionality

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -11,85 +12,75 @@ import {
+]
+
 /**
  * Observe results of querying the prompts from the Prompt Library, (deprecated) built-in commands,
```
Given the recent moves to observables, "observe" isn't a great verb choice. How about "Merges results of ..."

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -11,85 +12,75 @@ import {
-    )
+    const [customPrompts, isUnifiedPromptsEnabled] = await Promise.all([
+        fetchCustomPrompts(queryLower, signal),
+        featureFlagProvider.evaluateFeatureFlagEphemerally(FeatureFlag.CodyUnifiedPrompts),
```
Should this in fact *be* an observable and be polling the custom prompt fetch and observing the feature flag?

vovakulikov replied:
True, this is, in fact, a problem when you switch accounts, but your list of prompts is still the same until you refresh the prompt tab. 

I don't want to fix it in this PR because I expect some other thing may be broken. But good follow up

dominiccooney replied:
Consider putting a TODO to that effect, and linking to a Linear issue, so we have something to de-dup the QA bug reports against.

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -11,85 +12,75 @@ import {
-    ].filter(command => (isUnifiedPromptsEnabled ? { ...command, mode: 'ask' } : command))
+    const codyCommands = getCodyCommandList()
+    const allCommands: CommandAction[] = !clientCapabilities().isCodyWeb
+        ? // Ignore commands since with unified prompts vital commands will be replaced by out-of-box
```
Word choice, "vital" -> "essential"

Something weird with "unified prompts" too, is that possessive like "unified prompts' essential commands"?

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -11,85 +12,75 @@ import {
+                  ...c,
+                  actionType: 'command',
+              }))
+        : // Ignore any commands for Cody Web since none commands are supported there
```
"none commands are supported there" -> "no commands are supported there", or, "commands are not supported there"

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -98,20 +89,51 @@ export async function mergedPromptsAndLegacyCommands(
     )
 
+    const actions =
+        customPrompts !== 'unsupported' ? [...customPrompts, ...matchingCommands] : matchingCommands
```
Simplify `a !== b ? c : d` to `a === b ? d : c`.

Consider eliminating the stuttering of `matchingCommands` by floating the ternary inside the splice with `[...(x ? y : []), ...z]`

taiyab commented:
This is looking great! Good job @vovakulikov and @jasonhawkharris.

- Chat help should use the updated design from Jason's PR.
- Prompt/command labels should be slightly less bold
- We can knock down the icon sizes for prompts/commands down a bit (if they're 20x20, make them 16x16)
- If we're able to do some sort of sorting by last used, we should use that on the dropdown list from the prompts menu, and not the welcome area prompts (which should be sorted by last created by org (and published, not drafted)), if possible.
- I should not be able to see other peoples' drafted prompts anywhere, at all
- I should be able to see my drafted prompts in the library and last used dropdown only
- Edit Code needs to always exist on the upfront welcome area

dominiccooney commented on vscode/src/prompts/prompts.ts:
```diff
@@ -98,20 +89,51 @@ export async function mergedPromptsAndLegacyCommands(
+        }
+
+        const errorMessage = isErrorLike(error) ? error.message : String(error)
+        if (errorMessage.startsWith(`Cannot query field "prompts"`)) {
```
FYI when adding context filters we found some old Sourcegraph instances returned an empty string in some error cases. If this 'unsupported' state is important, you may want to test against old instances.

vovakulikov replied:
I don't think this is very important; worst case scenario, we don't show warning text that you're on an old instance. I checked it with 5.5 locally and it worked, but yeah I've seen before that errors sometimes are intercepted by something else on our backend

dominiccooney replied:
If it is not very important, do we need this code at all? Put a TODO on it to remove it when the support window for 5.5 closes?

dominiccooney commented on vscode/webviews/components/hooks.ts:
```diff
@@ -0,0 +1,22 @@
+
+export function useLocalStorage<T>(
+    key: string,
+    defaultValue?: T
```
Would be neat if there was a way callers providing a default were not punished with the `T | undefined` type.

vovakulikov replied:
I have as I think (hehe maybe not) a better implementation of this, will bring it in follow up PR.

dominiccooney commented on vscode/webviews/components/promptList/ActionItem.tsx:
```diff
@@ -0,0 +1,139 @@
+                </div>
+
+                <span className={styles.promptDescription}>
+                    {prompt.description ?? 'No description provided'}
```
Personally I dislike defaults like this because it is so much visual junk if people aren't providing descriptions... But if this is the design, so be it.

vovakulikov replied:
cc @taiyab it's your call

dominiccooney commented on vscode/webviews/components/promptList/PromptList.module.css:
```diff
@@ -0,0 +1,40 @@
+
+.list {
+    outline: none;
+    background-color: transparent !important;
```
What's going on here that we need to reach for `!important`?

vovakulikov replied:
Just a left over, thanks for spotting

dominiccooney commented on vscode/webviews/components/promptList/PromptList.module.css:
```diff
@@ -0,0 +1,40 @@
+        color: var(--vscode-input-foreground);
+
+        &:focus {
+            //border-color: var(--vscode-focusBorder);
```
I guess //-style comments are a feature of our CSS processor or does this mean something? If not let's remove it.

vovakulikov replied:
Just missed this, removed

dominiccooney commented on vscode/webviews/components/promptList/PromptList.tsx:
```diff
@@ -66,68 +69,50 @@ export const PromptList: React.FunctionComponent<{
-                  ? { type: 'command', value: codyCommand }
-                  : undefined
-            if (!entry) {
+            if (!action || !result) {
```
Simplify `!a || !b` -> `!(a && b)` which is one less operator and more emphatic that both are truthy.

vovakulikov replied:
disagree about this one, maybe it's a habit but I found that negative and or operator is better whey you try to unwrap this, you basically read it from left to right when with `!(a && b)` you have to keep in mind `!` operator when you read a && b

dominiccooney commented on vscode/webviews/components/promptList/PromptList.tsx:
```diff
@@ -66,68 +69,50 @@ export const PromptList: React.FunctionComponent<{
-                    isCommand: codyCommand ? 1 : 0,
-                    isCommandBuiltin: codyCommand?.type === 'default' ? 1 : 0,
-                    isCommandCustom: codyCommand?.type !== 'default' ? 1 : 0,
+                    isPrompt: isPrompt ? 1 : 0,
```
Consider `Number(isPrompt)` unless tsc yells at you.

dominiccooney commented on vscode/webviews/components/promptList/PromptList.tsx:
```diff
@@ -66,68 +69,50 @@ export const PromptList: React.FunctionComponent<{
                 },
                 privateMetadata: {
-                    nameWithOwner: prompt ? prompt.nameWithOwner : undefined,
+                    nameWithOwner: isPrompt ? action.nameWithOwner : undefined,
```
Would it be better to omit the `nameWithOwner` field entirely if it is not a prompt?

dominiccooney replied:
You could just `...isPrompt ? { nameWithOwner: ... } : {}` or something.

Really depends what the Telemetry consumer wants to see. Just seems weird to log the unrelated property with `undefined`.

dominiccooney commented on vscode/webviews/components/promptList/PromptList.tsx:
```diff
@@ -139,166 +124,83 @@ export const PromptList: React.FunctionComponent<{
+    const actions = showFirstNItems ? sortedActions.slice(0, showFirstNItems) : sortedActions
+
+    const inputPaddingClass =
+        paddingLevels !== 'none' ? (paddingLevels === 'middle' ? '!tw-p-2' : '!tw-px-4 !tw-py-4') : ''
```
As mentioned above, I'd use === and reverse the conditions so the reader doesn't have to think through negations. Ditto below.

How much does the important, and burying spacing in the code here, hurt us later? Can we make our own tailwind class for these and apply something logical here instead?

Why do you `!tw-px-4 !tw-py-4` and not simply `!tw-p-4`?

vovakulikov replied:
> Why do you !tw-px-4 !tw-py-4 and not simply !tw-p-4?

No reasons just a mistake, fixed it here (originally I think it was `!tw-px-4 !tw-py-6`

> How much does the important, and burying spacing in the code here, hurt us later? Can we make our own tailwind class for these and apply something logical here instead?

Yeah I will re-write this anyway this week, so will skip this comment for now, later it will be just a specific css class, yeah

dominiccooney commented on vscode/webviews/components/promptList/PromptList.tsx:
```diff
@@ -309,123 +211,13 @@ export const PromptList: React.FunctionComponent<{
-}
+function getSortedActions(actions: Action[], lastUsedActions: Record<string, number>): Action[] {
+    return [...actions].sort((action1, action2) => {
+        const action1Key = action1.actionType === 'prompt' ? action1.id : action1.key
```
Is there a lodash sortBy or something we could use here to cut down the duplication?

vovakulikov replied:
different keys on union type makes lodash `sortBy` a bit more tricky when I want it to be.

dominiccooney replied:
Isn't there a `sortBy` that takes a function that "projects" the value you want to sort on?

vovakulikov replied:
Yeah, but in this case if we use function it's not any different from native sort implementation like here

dominiccooney commented on vscode/webviews/components/promptList/fixtures.ts:
```diff
@@ -72,15 +77,17 @@ export function makePromptsAPIWithData(
-                        : data.prompts,
-                commands: data.commands?.filter(c => matchQuery(c.key)),
                 query,
+                isPromptsSupported: isPromptsSupported,
```
Just `isPromptsSupported` when the property name and variable name line up.

dominiccooney commented on vscode/webviews/components/promptList/utils.ts:
```diff
@@ -0,0 +1,5 @@
+import type { Action } from '@sourcegraph/cody-shared'
+
+export function commandRowValue(row: Action): string {
+    return row.actionType === 'prompt' ? `prompt-${row.id}` : `command-${row.key}`
```
There's a few places where you do this switch on `actionType` and consult `id` or `key`. Why not write a helper function and remove all the conditionals from those places?

dominiccooney commented on vscode/webviews/prompts/PromptsTab.tsx:
```diff
@@ -1,73 +1,76 @@
+
+    return (action: Action, setView: (view: View) => void) => {
+        try {
+            const actionKey = action.actionType === 'prompt' ? action.id : action.key
```
...places which could call that helper and be more readable++

dominiccooney reviewed (approved):
Looks good; many comments to simplify the code inline. I have generally commented the first time something occurs.

I am surprised/alarmed by the amount of `!` in the CSS, how does maintenance go over time with that?

vovakulikov commented:
> Chat help should use the updated design from Jason's PR.

Not a part of this PR will do as follow-up, or @jasonhawkharris will merge it as part of his PR

> Prompt/command labels should be slightly less bold, even tone of the color a bit like from Jason's PR

Set it to 500 now, let me know if it look ok.

> We can knock down the icon/avatar sizes for prompts/commands down a bit (if they're 20x20, make them 16x16)

16px seemed a bit off IMO. I reduced size down to 20 and 22 on the icon, should be ok now, also let me know

> If we're able to do some sort of sorting by last used, we should use that on the dropdown list from the prompts menu, and not the welcome area prompts (which should be sorted by last created by org (and published, not drafted)), if possible.

Ok, I turned off smart sorting on the prompt tab, but the welcome page seems like a good spot to have this functionality. This is based on what @rrhyne highly suggested last week and I agree with him here, any concerns about welcome page with sorting on? 

> I should not be able to see other peoples' drafted prompts anywhere, at all (@aramaraju only exception to this might be an admin?)

The problem here is that there is no API to turn off drafts from other users and keep your created drafts. I included drafts from everyone for now, but would be happy to turn off drafts completely if it's better (while we don't have behaviour about viewing only your drafts) 

> Edit Code needs to always exist on the upfront welcome area

Fixed, wonder if we need to show 5 items on the welcome area, because edit command will always be taking one spot from the list

> We can change the "Create Prompt' and "Prompt Library" links on the Prompts tab submenu to: "Create" and "Manage".

Would stick for now with prompt suffix, because we still have commands in place, Create and manage are too generic at the moment

> The "Find Code Smells" prompt should use the text-search icon from Lucide 

Fixed

taiyab commented:
> > Chat help should use the updated design from Jason's PR.
> 
> Not a part of this PR will do as follow-up, or @jasonhawkharris will merge it as part of his PR
> 
Cool.

> > Prompt/command labels should be slightly less bold, even tone of the color a bit like from Jason's PR
> 
> Set it to 500 now, let me know if it look ok.
> 

Looks much better 👍 

> > We can knock down the icon/avatar sizes for prompts/commands down a bit (if they're 20x20, make them 16x16)
> 
> 16px seemed a bit off IMO. I reduced size down to 20 and 22 on the icon, should be ok now, also let me know
> 

Feel free to knock down the avatars too. They should all be the same size and aligned.

> > If we're able to do some sort of sorting by last used, we should use that on the dropdown list from the prompts menu, and not the welcome area prompts (which should be sorted by last created by org (and published, not drafted)), if possible.
> 
> Ok, I turned off smart sorting on the prompt tab, but the welcome page seems like a good spot to have this functionality. This is based on what @rrhyne highly suggested last week and I agree with him here, any concerns about welcome page with sorting on?
> 

If we can only have ordered by recently used in the dropdown, that would be ideal. Welcome area needs to be reserved for discoverability of new prompts (not accessing ones that you're already using).

> > I should not be able to see other peoples' drafted prompts anywhere, at all (@aramaraju only exception to this might be an admin?)
> 
> The problem here is that there is no API to turn off drafts from other users and keep your created drafts. I included drafts from everyone for now, but would be happy to turn off drafts completely if it's better (while we don't have behaviour about viewing only your drafts)
> 

I'm happy to switch them all off in terms of visibility for now -- I assume we can keep them visible on the web to compensate, and so I can access drafted prompts whilst I'm drafting them via the web?

> > Edit Code needs to always exist on the upfront welcome area
> 
> Fixed, wonder if we need to show 5 items on the welcome area, because edit command will always be taking one spot from the list
> 

That would be OK for now. FYI, I'm not seeing Edit Code option when checking out your branch locally and authed to S2.

> > We can change the "Create Prompt' and "Prompt Library" links on the Prompts tab submenu to: "Create" and "Manage".
> 
> Would stick for now with prompt suffix, because we still have commands in place, Create and manage are too generic at the moment
> 

That's fine.

> > The "Find Code Smells" prompt should use the text-search icon from Lucide
> 
> Fixed

Sweet.

taiyab commented:
cc @danielmarquespt @taylorsperry for visibility

taiyab commented:
We can change the "Prompt Library" link at the bottom of the welcome area list to "All prompts"

thenamankumar commented on vscode/webviews/tabs/TabsBar.tsx:
```diff
@@ -285,15 +291,20 @@ const TabButton = forwardRef<HTMLButtonElement, TabButtonProps>((props, ref) =>
         'data-testid': dataTestId,
     } = props
 
+    const Component = uri ? 'a' : 'button'
```
better just render ahref and button separately with only 1 `uri ?` condition at top. Rather than merging them into a single component and then have multiple `uri?` in props.
